src/System.Slices/System/Buffers/Memory.cs

```diff
@@ -0,0 +1,132 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System;
+using System.Runtime;
+
+namespace System.Buffers
+{
+    /// <summary>
+    /// Represents a Span<byte> factory</byte>
+    /// </summary>
+    /// <remarks>
+    /// This struct is not safe for multithreaded access, i.e. it's subject to struct tearing that can result in unsafe memory access. 
+    /// </remarks>
+    public struct Memory<T>
+    {
+        public static unsafe Memory<T> Empty = new Memory<T>(null, 0, 0);
+
+        private readonly T[] _array;
+        private readonly int _offset;
+        private readonly unsafe void* _memory;
+        private readonly int _memoryLength;
+
+        public unsafe Memory(void* pointer, int length)
+        {
+            unsafe
+            {
+                _memory = pointer;
+            }
+
+            _array = null;
+            _offset = 0;
+            _memoryLength = length;
+        }
+
+        public unsafe Memory(T[] array) : this(array, 0, array.Length)
+        {
+        }
+
+        public unsafe Memory(T[] array, int offset, int length, bool isPinned = false)
```

Comment:
The default pool implementation in channels allocates a pre-pinned managed array. https://github.com/davidfowl/Channels/blob/master/src/Channels/MemoryPoolBlock.cs#L26

The same pool can be used for the managed socket API: https://github.com/davidfowl/Channels/blob/master/src/Channels.Networking.Sockets/SocketConnection.cs#L317

OR the libuv API without tweaks:
https://github.com/davidfowl/Channels/blob/master/src/Channels.Networking.Libuv/UvTcpConnection.cs#L199-L200

Now you could argue those 2 systems should just use different pools but if the user did all the hard work to pre-pinn the managed buffer passed in the both the unsafe pointer and the underlying array should be accessible.